### PR TITLE
[Merged by Bors] - Add AND/OR combinators for run conditions

### LINKED ISSUE #7584: Add OR and AND combinators for system run conditions
> I suspected that the only way to make an OR-combinator might be using a custom System impl, and now you confirmed it. So, I think it's out of scope for this PR.

Agreed. I'd like both OR and AND combinators (to follow-up #7559), but that's out of scope here.

_Originally posted by @alice-i-cecile in https://github.com/bevyengine/bevy/issues/7579#issuecomment-1424683773_

### PR DESCRIPTION
**Pull request by JoJoJet**
# Objective

Fix #7584.

## Solution

Add an abstraction for creating custom system combinators with minimal boilerplate. Use this to implement AND/OR combinators. Use this to simplify the implementation of `PipeSystem`.

## Example

Feel free to bikeshed on the syntax.

I chose the names `and_then`/`or_else` to emphasize the fact that these short-circuit, while I chose method syntax to empasize that the arguments are *not* treated equally.

```rust
app.add_systems((
    my_system.run_if(resource_exists::<R>().and_then(resource_equals(R(0)))),
    our_system.run_if(resource_exists::<R>().or_else(resource_exists::<S>())),
));
```

---

## Todo

- [ ] Decide on a syntax
- [x] Write docs
- [x] Write tests

## Changelog

+ Added the extension methods `.and_then(...)` and `.or_else(...)` to run conditions, which allows combining run conditions with short-circuiting behavior.
+ Added the trait `Combine`, which can be used with the new `CombinatorSystem` to create system combinators with custom behavior.

### CODE REVIEW

**Comment by JoJoJet** [comment]
@alice-i-cecile @inodentry I'd appreciate your opinions on the syntax used here :)

File: crates/bevy_ecs/src/schedule/condition.rs [hunk 1]
```diff
@@ -1,12 +1,31 @@
...
 
 pub type BoxedCondition = BoxedSystem<(), bool>;
 
 /// A system that determines if one or more scheduled systems should run.
 ///
 /// Implemented for functions and closures that convert into [`System<In=(), Out=bool>`](crate::system::System)
 /// with [read-only](crate::system::ReadOnlySystemParam) parameters.
-pub trait Condition<Params>: sealed::Condition<Params> {}
+pub trait Condition<Params>: sealed::Condition<Params> {
+    fn and_then<P, C: Condition<P>>(self, and_then: C) -> AndThen<Self::System, C::System> {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule/condition.rs:15]: Doc strings and doc tests here :)

I really like that these are methods though!

File: crates/bevy_ecs/src/schedule/condition.rs [hunk 2]
```diff
@@ -142,3 +161,69 @@ pub mod common_conditions {
...
+
+pub type OrElse<A, B> = CombinatorSystem<OrElseMarker, A, B>;
+
+#[doc(hidden)]
+pub struct AndThenMarker;
+
+impl<In, A, B> Combine<A, B> for AndThenMarker
+where
+    In: Copy,
+    A: System<In = In, Out = bool>,
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule/condition.rs:175]: Should this be restricted to ReadOnlySystem?

File: crates/bevy_ecs/src/system/combinator.rs [hunk 3]
```diff
@@ -0,0 +1,147 @@
+use std::{borrow::Cow, marker::PhantomData};
+
+use crate::{
+    archetype::ArchetypeComponentId, component::ComponentId, prelude::World, query::Access,
+};
+
+use super::{ReadOnlySystem, System};
+
+pub trait Combine<A: System, B: System> {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/combinator.rs:9]: Doc strings.

File: crates/bevy_ecs/src/system/combinator.rs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        self.a.get_last_change_tick()
+    }
+
+    fn set_last_change_tick(&mut self, last_change_tick: u32) {
+        self.a.set_last_change_tick(last_change_tick);
+        self.b.set_last_change_tick(last_change_tick);
+    }
+}
+
+/// SAFETY: Both systems are read-only, so the system combining them will only read from the world.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/combinator.rs:140]:
```suggestion
/// SAFETY: Both systems are read-only, so any system created by combining them will only read from the world.
```

**Comment by alice-i-cecile** [comment]
Milestone placement is "we should make a call on this", not "this must be included". I think there's a lot of value here, but it's also a little tricky.

File: crates/bevy_ecs/src/schedule/condition.rs (hunk 2 above)
Comment (JoJoJet) [reply crates/bevy_ecs/src/schedule/condition.rs:175]: Not necessary, `CombinatorSystem<A, B, AndThenMarker>` only implements `ReadOnlySystem` if `A` and `B` do, so the trait bounds "just work" as is.

**Comment by DJMcNab** [comment]
Why doesn't `pipe` use this combinator pattern?

I'd like us to avoid proliferating unsafe system implementations.

Two? is already too many, can we avoid it growing to three?

**Comment by JoJoJet** [comment]
> Why doesn't pipe use this combinator pattern?

It does.

**Comment by DJMcNab** [comment]
Ah, the dangers of trying to do code review on my phone. My apologies.

**Review by DJMcNab — COMMENTED** [review]
I'm not convinced by the safety comment on the calls to run_unsafe. They seem a bit blank.

(I also believe that they're not sound)

I can't say I especially like us stacking more content into `System` impls rather than just using proper systems. However, until Rust catches up, that's inevitable.

I also believe this new system is feasibly isomorphic to `ChainSystem`, but I also think the ergonomics are probably better.

**Comment by JoJoJet** [comment]
> (I also believe that they're not sound)

It's unsound in the same way that every other `System` impl is (has no guarantees that it's running on the correct `World`). I've been meaning to open a PR to address this.

File: crates/bevy_ecs/src/system/combinator.rs [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+/// # schedule.run(&mut world);
+/// # // Only the second condition passes, so the system runs.
+/// # assert!(world.resource::<RanFlag>().0);
+/// # world.resource_mut::<RanFlag>().0 = false;
+/// ```
+pub trait Combine<A: System, B: System> {
+    type In;
+    type Out;
+
+    fn combine(
```
[the remark this answers is not in the document]
Comment (JoJoJet) [reply crates/bevy_ecs/src/system/combinator.rs:85]: I ended up rewriting this trait and the way it's used, so this suggestion doesn't apply anymore. I also added comments for the relevant safety invariants, which should hopefully be more convincing now.

**Comment by JoJoJet** [comment]
I've added documentation and examples, simplified the `Combine` trait, and documented the previously-implicit safety invariants in `impl System for CombinatorSystem`.

File: crates/bevy_ecs/src/system/combinator.rs [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+        input: Self::In,
+        a: impl FnOnce(A::In) -> A::Out,
+        b: impl FnOnce(B::In) -> B::Out,
+    ) -> Self::Out;
+}
+
+/// A [`System`] defined by combining two other systems.
+/// The behavior of this combinator is specified by implementing the [`Combine`] trait.
+/// For a full usage example, see the docs for `Combine`.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/combinator.rs:101]:
```suggestion
/// For a full usage example, see the docs for [`Combine`].
```

**Review by alice-i-cecile — APPROVED** [review]
Very nice to see this.

**Comment by Shatur** [comment]
> I chose the names and_then/or_else to emphasize the fact that these short-circuit, while I chose method syntax to empasize that the arguments are not treated equally.

Maybe use `and` and `or`? Shorter and a little more readable as to me.

**Comment by JoJoJet** [comment]
Personally I prefer `and_then`/`or_else` for the reasons I mentioned in the PR desc -- I think that making the behavior clear is more important than making the name short. I'm willing to change it if there's a consensus for those names (or any other) though, of course.

**Comment by alice-i-cecile** [comment]
I prefer `and_then`/`or_else` as well :)

**Comment by BlackPhlox** [comment]
Could you add an example of usage under `examples/ecs`? There might already be a suitable .rs file

**Comment by JoJoJet** [comment]
From what I can tell, none of our examples use run conditions at all (except for `.on_update()`). We should address this, but IMO there should be a separate issue for this.

**Comment by Shatur** [comment]
Will it be possible to combine multiple conditions with this syntax?..

For example something like `A && B || C && D`?
I think it's very useful if you have substaces. For example, in my game I run `PlacingObjectPlugin::movement_system` in `GameState::City` with substate `CityMode::Objects` **and** in `GameState::Family` with substate `FamilyMode::Building`. Currently I have to add my system twice ([1](https://github.com/dollisgame/dollis/blob/98ddab5d74d25109f9b022e4d0a14381ffba1bc9/src/core/object/placing_object.rs#L49-L52), [2](https://github.com/dollisgame/dollis/blob/98ddab5d74d25109f9b022e4d0a14381ffba1bc9/src/core/object/placing_object.rs#L94-L97), the game currenly uses loopless, but will be similar with stageless).

**Comment by JoJoJet** [comment]
Yes, that should be possible. `and_then` and `or_else` return run conditions, so you should be able to keep adding as many combinators as you want.

**Comment by alice-i-cecile** [comment]
Can you add these to the example from #7652 once it's merged?

**Comment by JoJoJet** [comment]
I've added `and_then` to the example, but I'm really struggling to think of a way to show off `or_else` that isn't extremely contrived.

**Review by alice-i-cecile — APPROVED** [review]
Yeah, let's go with this: it'll hint to users that they should ask about the OR variant too.

**Comment by inodentry** [comment]
The most common use case for OR-combining run conditions that I've seen is to make a system run in multiple states (or other state-like resource type).

For example:

```rust
enum GameMode {
    Singleplayer,
    Multiplayer,
    Tutorial,
}

app.add_system(move_player_units.run_if(
    resource_equals(GameMode::Singleplayer)
        .or_else(resource_equals(GameMode::Multiplayer))
));
app.add_system(move_ai_units.run_if(
    resource_equals(GameMode::Singleplayer)
        .or_else(resource_equals(GameMode::Tutorial))
));
```

You could come up with something along those lines, for the example. It does not seem "contrived" to me. Very much a real-world use-case.

**Review by inodentry — APPROVED** [review]
I like this. I like the syntax, and also the implementation. Read through the code, didn't notice anything wrong. The Combine trait is clever, but useful. :)

The documentation is adequate.

**Pull request closed without merging**